{"expireTime":9007200904342771000,"key":"transformer-remark-markdown-html-41f17257f17af042a88c789503cf072e--","val":"<h1>Async functions:</h1>\n<h2>Asynchronous JavaScript</h2>\n<p>Let's start with the <code>async</code> keyword. It can be placed before a function, like this:</p>\n<pre><code class=\"language-js\">async function f() {\n    return 1;\n}\n</code></pre>\n<p>The word \"async\" before a function means one simple thing: a function always returns a promise. Other values are wrapped in a resolved promise automatically.</p>\n<p>For instance, this function returns a resolved promise with the result of <code>1</code>; let's test it:</p>\n<pre><code class=\"language-js\">async function f() {\n    return 1;\n}\n\nf().then(alert); // 1\n</code></pre>\n<p>...We could explicitly return a promise, which would be the same:</p>\n<pre><code class=\"language-js\">async function f() {\n    return Promise.resolve(1);\n}\n\nf().then(alert); // 1\n</code></pre>\n<p>So, <code>async</code> ensures that the function returns a promise, and wraps non-promises in it. Simple enough, right? But not only that. There's another keyword, <code>await</code>, that works only inside <code>async</code> functions, and it's pretty cool.</p>\n<h2><a href=\"#await\">Await</a></h2>\n<p>The syntax:</p>\n<pre><code class=\"language-js\">// works only inside async functions\nlet value = await promise;\n</code></pre>\n<p>The keyword <code>await</code> makes JavaScript wait until that promise settles and returns its result.</p>\n<p>Here's an example with a promise that resolves in 1 second:</p>\n<pre><code class=\"language-js\">async function f() {\n    let promise = new Promise((resolve, reject) => {\n        setTimeout(() => resolve('done!'), 1000);\n    });\n\n    let result = await promise; // wait until the promise resolves (*)\n\n    alert(result); // \"done!\"\n}\n\nf();\n</code></pre>\n<p>The function execution \"pauses\" at the line <code>(*)</code> and resumes when the promise settles, with <code>result</code> becoming its result. So the code above shows \"done!\" in one second.</p>\n<p>Let's emphasize: <code>await</code> literally suspends the function execution until the promise settles, and then resumes it with the promise result. That doesn't cost any CPU resources, because the JavaScript engine can do other jobs in the meantime: execute other scripts, handle events, etc.</p>\n<p>It's just a more elegant syntax of getting the promise result than <code>promise.then</code>. And, it's easier to read and write.</p>\n<p>Can't use <code>await</code> in regular functions</p>\n<p>If we try to use <code>await</code> in a non-async function, there would be a syntax error:</p>\n<hr>\n<pre><code class=\"language-js\">function f() {\n    let promise = Promise.resolve(1);\n    let result = await promise; // Syntax error\n}\n</code></pre>\n<p>We may get this error if we forget to put <code>async</code> before a function. As stated earlier, <code>await</code> only works inside an <code>async</code> function.</p>\n<p>Let's take the <code>showAvatar()</code> example from the chapter <a href=\"promise-chaining\">Promises chaining</a> and rewrite it using <code>async/await</code>:</p>\n<ol>\n<li>We'll need to replace <code>.then</code> calls with <code>await</code>.</li>\n<li>Also we should make the function <code>async</code> for them to work.</li>\n</ol>\n<pre><code class=\"language-js\">async function showAvatar() {\n    // read our JSON\n    let response = await fetch('/article/promise-chaining/user.json');\n    let user = await response.json();\n\n    // read github user\n    let githubResponse = await fetch(`https://api.github.com/users/${user.name}`);\n    let githubUser = await githubResponse.json();\n\n    // show the avatar\n    let img = document.createElement('img');\n    img.src = githubUser.avatar_url;\n    img.className = 'promise-avatar-example';\n    document.body.append(img);\n\n    // wait 3 seconds\n    await new Promise((resolve, reject) => setTimeout(resolve, 3000));\n\n    img.remove();\n\n    return githubUser;\n}\n\nshowAvatar();\n</code></pre>\n<p>Pretty clean and easy to read, right? Much better than before.</p>\n<pre><code class=\"language-js\">await` won't work in the top-level code\n\nPeople who are just starting to use `await` tend to forget the fact that we can't use `await` in top-level code. For example, this will not work:\n\n```js\n// syntax error in top-level code\nlet response = await fetch('/article/promise-chaining/user.json');\nlet user = await response.json();\n</code></pre>\n<p>But we can wrap it into an anonymous async function, like this:</p>\n<pre><code class=\"language-js\">(async () => {\n  let response = await fetch('/article/promise-chaining/user.json');\n  let user = await response.json();\n  ...\n})();\n</code></pre>\n<p>P.S. New feature: starting from V8 engine version 8.9+, top-level await works in <a href=\"modules\">modules</a>.</p>\n<pre><code class=\"language-js\">await` accepts \"thenables\"\n\nLike `promise.then`, `await` allows us to use thenable objects (those with a callable `then` method). The idea is that a third-party object may not be a promise, but promise-compatible: if it supports `.then`, that's enough to use it with `await`.\n\nHere's a demo `Thenable` class; the `await` below accepts its instances:\n\n```js\nclass Thenable {\n  constructor(num) {\n    this.num = num;\n  }\n  then(resolve, reject) {\n    alert(resolve);\n    // resolve with this.num*2 after 1000ms\n    setTimeout(() => resolve(this.num * 2), 1000); // (*)\n  }\n}\n\nasync function f() {\n  // waits for 1 second, then result becomes 2\n  let result = await new Thenable(1);\n  alert(result);\n}\n\nf();\n</code></pre>\n<p>If <code>await</code> gets a non-promise object with <code>.then</code>, it calls that method providing the built-in functions <code>resolve</code> and <code>reject</code> as arguments (just as it does for a regular <code>Promise</code> executor). Then <code>await</code> waits until one of them is called (in the example above it happens in the line <code>(*)</code>) and then proceeds with the result.</p>\n<p>Async class methods</p>\n<p>To declare an async class method, just prepend it with <code>async</code>:</p>\n<pre><code class=\"language-js\">class Waiter {\n    async wait() {\n        return await Promise.resolve(1);\n    }\n}\n\nnew Waiter().wait().then(alert); // 1 (this is the same as (result => alert(result)))\n</code></pre>\n<p>The meaning is the same: it ensures that the returned value is a promise and enables <code>await</code>.</p>\n<h2>[Error handling]</h2>\n<p>If a promise resolves normally, then <code>await promise</code> returns the result. But in the case of a rejection, it throws the error, just as if there were a <code>throw</code> statement at that line.</p>\n<p>This code:</p>\n<pre><code class=\"language-js\">async function f() {\n    await Promise.reject(new Error('Whoops!'));\n}\n</code></pre>\n<p>...is the same as this:</p>\n<pre><code class=\"language-js\">async function f() {\n    throw new Error('Whoops!');\n}\n</code></pre>\n<p>In real situations, the promise may take some time before it rejects. In that case there will be a delay before <code>await</code> throws an error.</p>\n<p>We can catch that error using <code>try..catch</code>, the same way as a regular <code>throw</code>:</p>\n<pre><code class=\"language-js\">async function f() {\n    try {\n        let response = await fetch('http://no-such-url');\n    } catch (err) {\n        alert(err); // TypeError: failed to fetch\n    }\n}\n\nf();\n</code></pre>\n<p>In the case of an error, the control jumps to the <code>catch</code> block. We can also wrap multiple lines:</p>\n<pre><code class=\"language-js\">async function f() {\n    try {\n        let response = await fetch('/no-user-here');\n        let user = await response.json();\n    } catch (err) {\n        // catches errors both in fetch and response.json\n        alert(err);\n    }\n}\n\nf();\n</code></pre>\n<p>If we don't have <code>try..catch</code>, then the promise generated by the call of the async function <code>f()</code> becomes rejected. We can append <code>.catch</code> to handle it:</p>\n<pre><code class=\"language-js\">async function f() {\n    let response = await fetch('http://no-such-url');\n}\n\n// f() becomes a rejected promise\nf().catch(alert); // TypeError: failed to fetch // (*)\n</code></pre>\n<p>If we forget to add <code>.catch</code> there, then we get an unhandled promise error (viewable in the console). We can catch such errors using a global <code>unhandledrejection</code> event handler as described in the chapter <a href=\"promise-error-handling\">Error handling with promises</a>.</p>\n<pre><code class=\"language-js\">async/await` and `promise.then/catch\n</code></pre>\n<p>When we use <code>async/await</code>, we rarely need <code>.then</code>, because <code>await</code> handles the waiting for us. And we can use a regular <code>try..catch</code> instead of <code>.catch</code>. That's usually (but not always) more convenient.</p>\n<p>But at the top level of the code, when we're outside any <code>async</code> function, we're syntactically unable to use <code>await</code>, so it's a normal practice to add <code>.then/catch</code> to handle the final result or falling-through error, like in the line <code>(*)</code> of the example above.</p>\n<pre><code class=\"language-js\">async/await` works well with `Promise.all\n</code></pre>\n<p>When we need to wait for multiple promises, we can wrap them in <code>Promise.all</code> and then <code>await</code>:</p>\n<pre><code class=\"language-js\">// wait for the array of results\nlet results = await Promise.all([\n  fetch(url1),\n  fetch(url2),\n  ...\n]);\n</code></pre>\n<p>In the case of an error, it propagates as usual, from the failed promise to <code>Promise.all</code>, and then becomes an exception that we can catch using <code>try..catch</code> around the call.</p>\n<h2><a href=\"#summary\">Summary</a></h2>\n<p>The <code>async</code> keyword before a function has two effects:</p>\n<ol>\n<li>Makes it always return a promise.</li>\n<li>Allows <code>await</code> to be used in it.</li>\n</ol>\n<p>The <code>await</code> keyword before a promise makes JavaScript wait until that promise settles, and then:</p>\n<ol>\n<li>If it's an error, the exception is generated --- same as if <code>throw error</code> were called at that very place.</li>\n<li>Otherwise, it returns the result.</li>\n</ol>\n<p>Together they provide a great framework to write asynchronous code that is easy to both read and write.</p>\n<p>With <code>async/await</code> we rarely need to write <code>promise.then/catch</code>, but we still shouldn't forget that they are based on promises, because sometimes (e.g. in the outermost scope) we have to use these methods. Also <code>Promise.all</code> is nice when we are waiting for many tasks simultaneously.</p>\n<h2>Cleaning up your asynchronous code with <code>await</code></h2>\n<p>REPLs have traditionally had a difficult time allowing you to interact with asynchronous code since they encourage a coding style where you evaluate expressions and use those results in the next field. But if you are using promises or callbacks, this breaks down because these results exist only in the callback, not the next line:</p>\n<p>In RunKit, you can use top-level <code>await</code> to get the same behavior of synchronous code.</p>\n<p>Now we can treat this code as synchronous, despite the fact that the code is still executing asynchronously.</p>\n<p>Let's look how. It helps to have a more complex example, where we need to do a few asynchronous operations in sequence. You can see how <code>await</code>, promises and callbacks achieve the same results, but the <code>await</code> style works better in a REPL:</p>\n<ul>\n<li>await</li>\n<li>promises</li>\n<li>callbacks</li>\n</ul>\n<p>Here, we use <code>await</code> on lines 4 and 8, and the results from each request remain in scope.</p>\n<p>Remember, <code>await</code> expects a <code>promise</code> so you can either write your own or use one of the many libraries that natively supports promises, and npm is full of packages that add promise support to existing libraries. Here are a few of our favorites:</p>\n<ul>\n<li><a href=\"https://npm.runkit.com/fs-promise\">fs-promise</a> - promise based filesystem api</li>\n<li><a href=\"https://npm.runkit.com/request-promise\">request-promise</a> - a wrapper around \"request\" for http stuff</li>\n<li><a href=\"https://npm.runkit.com/glob-promise\">glob-promise</a> - glob style filesystem queries</li>\n<li><a href=\"https://npm.runkit.com/bluebird\">bluebird</a> - general promise library with lots of utilities</li>\n</ul>\n<h3>Further Reading</h3>\n<ul>\n<li><a href=\"http://rossboucher.com/await\">ES7 Async/Await presented at Brookyln.js</a></li>\n<li><a href=\"https://babeljs.io/\">Babel.js</a></li>\n<li><a href=\"https://github.com/lukehoban/ecmascript-asyncawait/\">ECMAScript's Proposal for async/await</a></li>\n<li><a href=\"https://esdiscuss.org/notes/2014-01-30#async-await\">ES Meeting Notes discussing on async/await</a></li>\n<li><a href=\"https://esdiscuss.org/topic/does-async-await-solve-a-real-problem\">Does async/await Solve a Real Problem?</a></li>\n<li><a href=\"https://thomashunter.name/blog/the-long-road-to-asyncawait-in-javascript/\">The Long Road to async/await in JavaScript</a></li>\n<li><a href=\"https://www.twilio.com/blog/2015/10/asyncawait-the-hero-javascript-deserved.html\">async/await: The Hero JavaScript Deserved</a></li>\n</ul>"}